networkx/generators/geometric.py

```diff
@@ -156,9 +163,148 @@ def random_geometric_graph(n, radius, dim=2, pos=None, p=2):
+    Returns
+    -------
+    Graph
+        A soft random geometric graph, undirected and without self-loops.
+        Each node has a node attribute ``'pos'`` that stores the
+        position of that node in Euclidean space as provided by the
+        ``pos`` keyword argument or, if ``pos`` was not provided, as
+        generated by this function.
+
+    Examples
+    --------
+    Default Graph:
+
+    G = nx.soft_random_geometric_graph(n, 0.2)
+
+    Custom Graph:
+
+    Create a soft random geometric graph on 100 uniformly distributed nodes
+    where nodes are joined by an edge with probability computed from an exponential
+    distribution with rate parameter :math:`\lambda=1` if their Euclidean distance 
+    is at most 0.2::
+
+    Notes
+    -----
+    This uses a *k*-d tree to build the graph.
+
+    The `pos` keyword argument can be used to specify node positions so you
+    can create an arbitrary distribution and domain for positions.
+
+    For example, to use a 2D Gaussian distribution of node positions with mean
+    (0, 0) and standard deviation 2
+
+    The scipy.stats package can be used to define the probaility distribution
+    with the .pdf method used as `p_dist`.
+
+    ::
+
+    >>> import random
+    >>> import math
+    >>> n = 100
+    >>> p = {i: (random.gauss(0, 2), random.gauss(0, 2)) for i in range(n)}
+    >>> def p_dist(dist):
+    ...     return math.exp(-dist)
+    >>> G = nx.soft_random_geometric_graph(n, 0.2, pos=p, p_dist=p_dist)
+
+    References
+    ----------
+    .. [1] Penrose, Mathew D. "Connectivity of soft random geometric graphs."
+           The Annals of Applied Probability 26.2 (2016): 986-1028.
+       [2] scipy.stats - https://docs.scipy.org/doc/scipy/reference/tutorial/stats.html
+
+    """
+
+    n_name, nodes = n
+    G = nx.Graph()
+    G.name = 'soft_random_geometric_graph({}, {}, {})'.format(n, radius, dim)
+    G.add_nodes_from(nodes)
+    # If no positions are provided, choose uniformly random vectors in
+    # Euclidean space of the specified dimension.
+    if pos is None:
+        pos = {v: [random.random() for i in range(dim)] for v in nodes}
+    nx.set_node_attributes(G, 'pos', pos)
+
+    #if p_dist function not supplied define uniform connection probability function
+    #where all node pairs connect with constant probability = 0.5 at any distance
+    if p_dist is None:
+
+        def p_dist(dist):
+            return 0.5
+
+    def should_join(pair):
+        u, v = pair
+        u_pos, v_pos = pos[u], pos[v]
+        dist = (sum(abs(a - b) ** p for a, b in zip(u_pos, v_pos)))**(1/p)
+        #Check if dist is <= radius parameter. This check is redundant if scipy
+        #is availible and _fast_edges routine is used, but provides the check incase
+        #scipy is not availible and all edge combinations need to be checked
+        if dist <= radius:
+            return random.random() < p_dist(dist)
+        else:
+            return False
+
+    if _is_scipy_available:
```

Comment:
If scipy is available, we use the KDTree implementation to find the possible edges, and use that to add these possible edges with probability obtained from the supplied pdf. Note that this requires another iteration over the possible edge list and recomputing their distances, but my intuition is that this method will scale better than running a single iteration over all possible edge combinations. Might be a good idea to do some performance testing for the two possibilities though.